fix(core): components marked for traversal resets reactive context

kristilw opened the pull request:
when marked for traversal the reactive context has to be set to null to avoid inheriting the reactive context of the parent component

PR Closes https://github.com/angular/angular/issues/61662

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #61662


## What is the new behavior?
The embedded view updates its template when it depends on a signal which has changed its value.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

atscott reviewed (approved):
This is correct in isolation.

(TMI disclaimer) I'm still not sure about how I feel about the attached view case in general. It wasn't the intent that attaching a view would skip refreshing the component. Skipping it will result in its queries not being updated. E.g. if I add a test like this:


```
     @Directive({
          selector: '[step]'
        })
        class Step {}
    
        @Directive({selector: '[dataDirective]'})
        class DataDirective {
          readonly templateRef = inject(TemplateRef);
          readonly viewContainerRef = inject(ViewContainerRef);
    
          drawTemplate() {
            this.viewContainerRef.createEmbeddedView(this.templateRef, {$implicit: data});
          }
        }
    
        @Component({
          selector: 'child',
          template: '<ng-container *dataDirective="let data"><div step> {{data()}}</div></ng-container>',
          imports: [DataDirective, Step],
          changeDetection: ChangeDetectionStrategy.OnPush,
        })
        class ChildComponent {
          @ViewChildren(Step) steps?: QueryList<Step>;
          signalSteps = viewChildren(DataDirective);
        }
        ...
```

The length of `steps` and `signalSteps` should really both be 1 after you call `drawTemplate` and wait for stability, but it's not because view queries are updated at the end of refreshing the component view: https://github.com/angular/angular/blob/c11527fc68965fee9799cc4d1f6f0ea33fce98ce/packages/core/src/render3/instructions/change_detection.ts#L299-L302

This is one of the reasons we ensure that embedded views share consumers with their parent component and why the temporary consumer marks the component view for refresh. The "correct" thing to do for this case would be to ensure that the component gets refreshed when an embedded view is attached but that comes with its own difficulties. There are similar problems with the old style queries that aren't necessarily solvable and we can kind of shrug and say to use signal queries to fix those problems.

This issue already exists and again the change here is still correct and doesn't introduce any regression. But I think it's worth noting that this "fixes" the issue in a way that's not necessarily directly addressing the problem with the attached view not marking its parent OnPush component for check.

kristilw committed:
refactor(core): update tests to use standalone

update tests to use standalone components for easier test setup

kristilw committed:
fix(core): components marked for traversal resets reactive context

when marked for traversal the reactive context has to be set to null to avoid inheriting the reactive context of the parent component

PR Closes #61662

kristilw commented on packages/core/src/render3/instructions/change_detection.ts:
```diff
@@ -490,16 +491,22 @@ function detectChangesInView(lView: LView, mode: ChangeDetectionMode) {
-    if (!isInCheckNoChangesPass) {
-      addAfterRenderSequencesForView(lView);
+    // Sets active consumer to null, ensuring that an improper reactive context is not inherited
+    const prevConsumer = setActiveConsumer(null);
```
Updated. Just like to add that `getActiveConsumer() === null` would already be the case if you take my original bug report and add OnPush change detection to the ParentComponent

kristilw commented on packages/core/test/change_detection_scheduler_spec.ts:
```diff
@@ -838,6 +840,51 @@ describe('Angular with scheduler and ZoneJS', () => {
     expect(fixture.nativeElement.innerText).toContain('new');
   });
 
+  it('updating signal inside an EmbeddedView in a child component with OnPush inside a parent component with Default CD', async () => {
```
I'm also not too happy with the long name of the test, and it does not even convey that the child component has to be only marked for traversal and not dirty. Not sure if best practice in this repo is to have a long test name or add a comment to the test. I'll take your suggestion If you have any.

kristilw replied:
If it was my repo I would rather have a longer test name. But I'm not deeply familiar with this repo, and I'm mostly concerned with fixing the bug, so I'll go with whatever makes you happy 😁 But I'll try to think of something

atscott replied:
It doesn't really matter. Long name is fine. I'm happy if you're happy.

JeanMeche commented:
Yes, the failed test was a flaky one.
